demoapp/demo_app_functions.py

```diff
@@ -450,4 +512,642 @@ def update_csv(bill_num: str, title: str, summarized_bill: str, category: str, t
+    )
+
+    # call the summary function
+    status_code, results = get_tags(bill_details)
+
+    # return response attribute of returned value
+    if status_code != 1: 
+        return {'status': status_code, 'tags': []}
+    else: 
+        return {'status': status_code, 'tags': results.response}
+
+def get_llm_call_type(bill_details: BillDetails) -> str:
+
+    """
+    This function calculates number of tokens and decides on weather to use RAG or not. It reutrns a string output
+    that specifies how to call the LLM. 
+
+    Args: 
+        bill_details (BillDetails): object consisting of bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        str: 'large' or 'small' depeneding upon token count
+
+    """
+
+    bill_text = getattr(bill_details, "bill_text")
+    bill_title = getattr(bill_details, "bill_title")
+    mgl_ref = getattr(bill_details, "mgl_ref")
+    committee_info = getattr(bill_details, "committee_info")
+    mgl_names = getattr(bill_details, "mgl_names")
+
+    num_tokens = count_tokens(bill_title, bill_text, mgl_ref, mgl_names, committee_info)
+    
+    return 'small' if num_tokens < MAX_TOKEN_LIMIT - 5000 else 'large'
+
+def get_category_tags(categories: List) -> List:
+
+    """
+    This function takes in list of categories and returns tags pertinant to that specifc categories only. 
+
+    Args: 
+
+        categories (List(str)): List of category strings.
+
+    Returns: 
+        List of all tags specific to those of categories.  
+    """
+
+    tags_tuple = itemgetter(*set.intersection(set(categories), set(new_categories_for_bill_list)))(new_tags_for_bill_dict)
+    
+    if isinstance(tags_tuple, list): return tags_tuple
+
+    category_tags = []
+    for cts in tags_tuple: category_tags += cts
+    return category_tags
+
+def get_summary(bill_details: BillDetails) -> tuple[int, LLMResults]:
+    '''
+    This function takes in bill details object (bill title, bill text and reference mgl section text) and summarizes the bill. 
+
+    Arguments: 
+
+    bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        A tuple of status_code and an LLMResults object containing query, response from the LLM
+        status_code can take these following values {1: Success, -1: Necessary details not found}
+
+    '''
+
+    if not all(hasattr(bill_details, attr) for attr in ("bill_text", 'bill_title', 'mgl_names', 'committee_info')): 
+        return -1, LLMResults()
+
+    set_my_llm_cache()
+    llm_call_type = get_llm_call_type(bill_details)
+
+    query = get_query_for_summarization(bill_details, llm_call_type)
+    return 1, call_llm(bill_details, query, llm_call_type)
+
+def get_tags(bill_details: BillDetails) -> tuple[int, LLMResults]:
+    '''
+    This function takes in bill details object (bill title, bill text and reference mgl section text) and tags the bill. 
```

Comment:
Can you add a sentence about how you do the tagging in two stages, first the categories and then the tags?